src/Compilers/Core/Portable/Symbols/Attributes/CommonAttributeData.cs

```diff
@@ -425,6 +426,20 @@ internal static void DecodeMethodImplAttribute<T, TAttributeSyntaxNode, TAttribu
                     messageProvider.ReportInvalidAttributeArgument(arguments.Diagnostics, arguments.AttributeSyntaxOpt, 0, attribute);
                     options = options & ~(MethodImplOptions)3;
                 }
+
+                // https://github.com/dotnet/roslyn/issues/79792: Use the real value when possible
+                const MethodImplOptions MethodImplOptionsAsync = (MethodImplOptions)0x2000;
+                if ((options & MethodImplOptionsAsync) != 0)
+                {
+                    // Error if [MethodImpl(MethodImplOptions.Async)] is used directly on a method
+                    // We give an exception to the AsyncHelpers special type, as it manually implements the pattern as part of the
+                    // runtime's async support
+                    if ((InternalSpecialType)appliedToSymbol.ExtendedSpecialType != InternalSpecialType.System_Runtime_CompilerServices_AsyncHelpers)
+                    {
+                        arguments.Diagnostics.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_MethodImplAttributeAsyncCannotBeUsed, arguments.AttributeSyntaxOpt.Location));
+                        options &= ~MethodImplOptionsAsync;
```

Comment:
does doing this actually matter?  seems slightly odd (to me at least) that we care about mutating the value once we've already given a full error.

Reply:
As a general compiler principle, if something like this is invalid, we don't then want to expose that to users that query for method impl options via our API.

Reply:
good to know.  Thanks!

Reply:
#resolved